bentoml/_internal/exportable.py

```diff
@@ -0,0 +1,232 @@
+            resource = parsedurl.resource
+            params = parsedurl.params
+            subpath = parsedurl.path
+
+        if protocol not in fs.opener.registry.protocols:
+            if protocol == "s3":
+                raise ValueError(
+                    "Tried to open an S3 url but the protocol is not registered; did you install fs-s3fs?"
+                )
+            else:
+                raise ValueError(
+                    f"Unknown or unsupported protocol {protocol}. Some supported protocols are ftp, s3, and osfs."
+                )
+
+        isOSPath = protocol in [
+            "osfs",
+            "userdata",
+            "userconf",
+            "sitedata",
+            "siteconf",
+            "usercache",
+            "userlog",
+        ]
+
+        isCompressedPath = protocol in ["tar", "zip"]
+
+        if output_format is not None:
+            if isCompressedPath:
+                raise ValueError(
+                    f"A {protocol} path was passed along with an output format ({output_format}); only pass one or the other."
+                )
+        else:
+            # try to guess output format if it hasn't been passed
+            output_format = self.guess_format(resource if subpath is None else subpath)
+
+        # use combine here instead of join because we don't want to fold subpath into path.
+        fullpath = resource if subpath is None else fs.path.combine(resource, subpath)
+
+        if fullpath == "" or (not output_format == "folder" and fullpath[-1] == "/"):
+            subpath = "" if subpath is None else subpath
+            # if the path has no filename or is empty, we generate a filename from the bento tag
+            subpath = fs.path.join(subpath, self._export_name())
+            if output_format in ["gz", "xz", "bz2", "tar", "zip"]:
+                subpath += "." + output_format
+        elif isOSPath:
+            try:
+                dirfs = fs.open_fs(path)
+            except fs.errors.CreateFailed:
+                pass
+            else:
+                if subpath is None or dirfs.isdir(subpath):
+                    subpath = "" if subpath is None else subpath
+                    # path exists and is a directory, we will create a file inside that directory
+                    subpath = fs.path.join(subpath, self._export_name)
+                    if output_format in ["gz", "xz", "bz2", "tar", "zip"]:
+                        subpath += "." + output_format
+
+        if subpath is None:
+            if isCompressedPath:
+                subpath = ""
+            else:
+                resource, subpath = fs.path.split(resource)
+
+        # as a special case, if the output format is the default we will always append the relevant
+        # extension if the output filename does not have it
+        if (
+            output_format == self._export_ext
+            and fs.path.splitext(subpath)[1] != "." + self._export_ext
+        ):
+            print(
```

Comment:
should we do `logger` instead of `print`?

Reply:
👍 done.